### DISCUSSION TITLE
import from node_modules

### DISCUSSION THREAD

**Discussion by darwin808** [vercel/next.js#36768] [Help]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

    Operating System:
      Platform: linux
      Arch: x64
      Version: #1 SMP PREEMPT Wed, 27 Apr 2022 20:56:11 +0000
    Binaries:
      Node: 14.17.6
      npm: 6.14.15
      Yarn: 1.22.18
      pnpm: N/A
    Relevant packages:
      next: 12.1.6
      react: 18.1.0
      react-dom: 18.1.0

### What browser are you using? (if relevant)

Chrome 100

### How are you deploying your application? (if relevant)

next start,

### Describe the Bug

Im trying to import a bundle.js from node_modules  to next js using a script tag 

`   <Script
               type="module"
               strategy="beforeInteractive"
               src="/node_modules/@myLib/build/bundle.js"
            />
`

and I get this error in client
`GET http://localhost:3000/node_modules/@mylibrary/build/bundle.js net::ERR_ABORTED 404 (Not Found)`

also tried doing the same in react and vite and its working fine 

### Expected Behavior

It should work because its working in vanilla react / vite react 

### To Reproduce

pick a library example jquery.min.js then put it inside the node_modules folder and then try to import it in the _document.js or _app.js

`
   <Script
               type="module"
               strategy="beforeInteractive"
               src="/node_modules/@myLib/build/bundle.js"
            />

`

**Comment by vadymshymko** [comment]
The content of the node_modules folder is not available in the browser. I think In order to make some files available, you can use the [copy-webpack-plugin[https://github.com/webpack-contrib/copy-webpack-plugin)  to copy the necessary files to the build output folder. 

In your next.condig.js:
```
const nextConfig = {
  webpack: (config) => {
    const customPlugins = [
      new CopyPlugin({
        patterns: [
          {
            from: 'node_modules/path/to/file',
            to: 'static/file',
          },
        ],
      }),
    ];

    config.plugins.push(...customPlugins);

    return config;
  },
};

module.exports = nextConfig;
```
somewhere in the code:
`<Script type="module" src="/_next/static/file" strategy="beforeInteractive" />`

**Reply by vadymshymko** [reply]
Upd. Oh, I think the solution presented by @icyJoseph is better. Try to use it)

**Reply by darwin808** [reply]
> The content of the node_modules folder is not available in the browser. I think In order to make some files available, you can use the [copy-webpack-plugin[https://github.com/webpack-contrib/copy-webpack-plugin) to copy the necessary files to the build output folder.
> 
> In your next.condig.js:
> 
> ```
> const nextConfig = {
>   webpack: (config) => {
>     const customPlugins = [
>       new CopyPlugin({
>         patterns: [
>           {
>             from: 'node_modules/path/to/file',
>             to: 'static/file',
>           },
>         ],
>       }),
>     ];
> 
>     config.plugins.push(...customPlugins);
> 
>     return config;
>   },
> };
> 
> module.exports = nextConfig;
> ```
> 
> somewhere in the code: `<Script type="module" src="/_next/static/file" strategy="beforeInteractive" />`

Tried this and it recreates the file inside the .next folder still cant be access from the script tag

> Hi,
> 
> If you just want to make a script available on the browser, simply import it on `_app.tsx`, but do remember that some modules do checks on the window/document, when they are loaded, so you might need to dynamically load the script, or at the very least do some window checks of your own.
> 
> For example to make `jQuery` globally available:
> 
> ```tsx
> import "../styles/globals.css";
> import type { AppProps } from "next/app";
> import jQuery from "jquery";
> 
> if (typeof window !== "undefined") {
>   window.jQuery = jQuery; // TypeScript will complain here. Use a global.d.ts file to define jQuery globally
>   // document is also available inside here.
> }
> 
> function MyApp({ Component, pageProps }: AppProps) {
>   return <Component {...pageProps} />;
> }
> 
> export default MyApp;
> ```
> 
> If the script/module is not needed for the initial load, then do a dynamic import inside a `useEffect`.
> 
> To the point of `next/script` itself, the docs say about the `src` prop:
> 
> > A path string specifying the URL of an external script. This can be either an absolute external URL or an internal path.
> 
> Perhaps internal path is misread here, but it refers to relative paths to the application, not the node project. For instance, if you have a script in the public folder you can load it by using `/script-name.js`.
> 
> To sum up, in order to make `node_modules` available, you don't need to use scripts, you need to explicitly make them available in the window, either statically or dynamically.

**Reply by darwin808** [reply]
also tried this its working and i make a folder in root example static/bundle.js , but still node_modules or .next folder is giving me 404 not found error I am using npm link thats why I need to access the node_modules for changes in dev.

**Reply by icyJoseph** [reply]
I think when linking, every time you make changes to your local package, you ought to force install again on the main project, when this is a separate project anyway. Otherwise, I think, even if you expose the package through a script, it'd still be the out of date.

I don't think you need to use the script tag at all though, importing the node_module at the top of `_app.tsx` should be enough. How is this not working out for you? Would it be possible to take a look at what you are trying to do?

---

### An alternative

Let us remember that, api routes can serve most types of content, we are limited by size rather, so we can serve `application/javascript` or its now deprecated pair `text/javascript` from there.

I've set up a project called `sum-script`:

```js
const createSum = () => {
  const loadTime = Date.now();

  return (a, b) => {
    const result = a + b;
    console.log("Sum lib created at:", loadTime);
    console.log(`${a} + ${b} = ${result}`);
    return result; 
  }
}

if (typeof window !== 'undefined') {
  window.sumLib = createSum();
}
```

Then, on another project, I install sum-script as a dependency via `yarn add file:/path/to/sum-script`.

Now I define this `api/route`:

```ts
import type { NextApiRequest, NextApiResponse } from "next";
import fs from "fs/promises";
import path from "path";

async function handler(req: NextApiRequest, res: NextApiResponse) {
  // important to use cwd, because API routes run in a special environment
  const script = await fs.readFile(
    path.join(process.cwd(), "node_modules", "sum-script", "index.js")
  ); // reads the script as text

  res.setHeader("content-type", "application/javascript");
  return res.send(script); // send it back to the client
}

export default handler;
```

And then I can just `next/script` like so:

```jsx
   <Script src="/api/hello"></Script>
```

And that loads the script. 

I am not sure if this would totally work when deploying though. Again, I'd go for importing the dependency on `_app.tsx`, rather.

**Comment by icyJoseph** [comment]
Hi,

If you just want to make a script available on the browser, simply import it on `_app.tsx`, but do remember that some modules do checks on the window/document, when they are loaded, so you might need to dynamically load the script, or at the very least do some window checks of your own.

For example to make `jQuery` globally available: 

```tsx
import "../styles/globals.css";
import type { AppProps } from "next/app";
import jQuery from "jquery";

if (typeof window !== "undefined") {
  window.jQuery = jQuery; // TypeScript will complain here. Use a global.d.ts file to define jQuery globally
  // document is also available inside here.
}

function MyApp({ Component, pageProps }: AppProps) {
  return <Component {...pageProps} />;
}

export default MyApp;
```

If the script/module is not needed for the initial load, then do a dynamic import inside a `useEffect`.

To the point of `next/script` itself, the docs say about the `src` prop:

> A path string specifying the URL of an external script. This can be either an absolute external URL or an internal path.

Perhaps internal path is misread here, but it refers to relative paths to the application, not the node project. For instance, if you have a script in the public folder you can load it by using `/script-name.js`.

To sum up, in order to make `node_modules` available, you don't need to use scripts, you need to explicitly make them available in the window, either statically or dynamically.
